### PR TITLE
Alkwa/upgrade to latest

### PR DESCRIPTION
**Pull request by alkwa-msft**
## Purpose

* ...

## Does this introduce a breaking change?

```
[ ] Yes
[ ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

* Test the code

```
```

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

File: Chat/ClientApp/src/App.tsx [hunk 1]
```diff
@@ -31,43 +31,19 @@ export default (): JSX.Element => {
     } else if (page === 'configuration') {
       return (
         <ConfigurationScreen
-          joinChatHandler={() => {
-            setPage('chat');
-          }}
-        />
+          endChatHandler={() => setPage('error')}
```
Comment (alkwa-msft) [inline Chat/ClientApp/src/App.tsx:34]: rename to kicked out handler or something else

File: Chat/ClientApp/src/components/ChatThread.tsx [hunk 2]
```diff
@@ -107,11 +110,8 @@ export default (props: ChatThreadProps): JSX.Element => {
   }, [numberOfMessagesToRender]);
 
   useEffect(() => {
-    if (
-      props.messages.length > 0 &&
-      props.messages[props.messages.length - 1].sender?.communicationUserId ===
-        props.user.identity
-    ) {
+    const user = props.messages && props.messages[props.messages.length - 1] && props.messages[props.messages.length - 1].sender;
+    if (props.messages.length > 0 && user && isUserMatchingIdentity(user, props.user.identity)) {
```
Comment (alkwa-msft) [inline Chat/ClientApp/src/components/ChatThread.tsx:114]: add a comment what is going on here

File: Chat/ClientApp/src/components/ConfigurationScreen.tsx [hunk 3]
```diff
@@ -37,7 +37,8 @@ import { MAXIMUM_LENGTH_OF_NAME } from '../constants';
 
 export interface ConfigurationScreenProps {
   joinChatHandler(): void;
-  setup(displayName: string, emoji: string, joinChatHandler: Function): void;
+  endChatHandler(): void;
```
Comment (alkwa-msft) [inline Chat/ClientApp/src/components/ConfigurationScreen.tsx:40]: similar to comment above its really that you are removed from the chat and you need to handle it

File: Chat/ClientApp/src/components/SidePanel.tsx [hunk 4]
```diff
@@ -29,67 +27,34 @@ import {
...
-      if (needToSet) {
-        setContosoUsers(users);
-      }
-    };
-    fetchData();
-  }, [users, threadMembers, setContosoUsers]);
-
-  useEffect(() => {
-    if (removeThreadMemberError) {
+    if (removeChatParticipantError) {
```
Comment (alkwa-msft) [inline Chat/ClientApp/src/components/SidePanel.tsx:51]: I question if we still need to handle removeChatParticipant in this component

File: Chat/ClientApp/src/containers/ConfigurationScreen.ts [hunk 5]
```diff
@@ -10,8 +10,8 @@ const mapStateToProps = (state: State) => ({
 });
 
 const mapDispatchToProps = (dispatch: any) => ({
-  setup: async (displayName: string, emoji: string, goToNextScreen: Function) => {
-    await dispatch(addUserToThread(displayName, emoji));
+  setup: async (displayName: string, emoji: string, endChatHandler: () => void) => {
```
Comment (alkwa-msft) [inline Chat/ClientApp/src/containers/ConfigurationScreen.ts:13]: kicked handler plz

File: Chat/ClientApp/src/containers/TypingIndicator.ts [hunk 6]
```diff
@@ -3,6 +3,14 @@ import { connect } from 'react-redux';
 import TypingIndicator from '../components/TypingIndicator';
 import { State } from '../core/reducers/index';
 import { PARTICIPANTS_THRESHOLD, MAXIMUM_LENGTH_OF_TYPING_USERS } from '../constants';
+import { ChatParticipant } from '@azure/communication-chat';
+import { CommunicationUserIdentifier } from '@azure/communication-common';
+
+type TypingUser = {
```
Comment (alkwa-msft) [inline Chat/ClientApp/src/containers/TypingIndicator.ts:9]: removed locally already

File: Chat/ClientApp/src/core/reducers/MessagesReducer.ts [hunk 7]
```diff
@@ -38,7 +47,7 @@ export const MessagesReducer = (state = initMessagesState, action: MessagesActio
         ...state,
         typingNotifications: {
           ...state.typingNotifications,
-          [action.id]: action.typingNotification
+          [action.id]: action.latestTypingNotification
```
Comment (alkwa-msft) [inline Chat/ClientApp/src/core/reducers/MessagesReducer.ts:50]: rename back to typingNotification to create less confusion

File: Chat/ClientApp/src/core/sideEffects.ts [hunk 8]
```diff
@@ -6,15 +6,10 @@ import {
...
-  TOO_MANY_REQUESTS_STATUS_CODE,
-  PRECONDITION_FAILED_STATUS_CODE,
   MAXIMUM_RETRY_COUNT,
-  COOL_PERIOD_THRESHOLD,
-  OK,
-  CREATED,
-  MULTI_STATUS
+  OK
 } from '../constants';
-import { setChatClient, setContosoUser, setContosoUserCoolPeriod } from './actions/ContosoClientAction';
```
Comment (alkwa-msft) [inline Chat/ClientApp/src/core/sideEffects.ts:17]: are we still using cooldown period

File: Chat/ClientApp/src/core/sideEffects.ts [hunk 9]
```diff
@@ -26,21 +21,22 @@ import {
...
-  GetChatMessageResponse
+  ChatParticipant,
 } from '@azure/communication-chat';
-import { AzureCommunicationUserCredential, RefreshOptions } from '@azure/communication-common';
+import { AzureCommunicationTokenCredential, CommunicationTokenRefreshOptions, CommunicationUserIdentifier } from '@azure/communication-common';
+import { ChatThreadPropertiesUpdatedEvent, ParticipantsAddedEvent, ParticipantsRemovedEvent } from '@azure/communication-signaling';
 
 // This function sets up the user to chat with the thread
-const addUserToThread = (displayName: string, emoji: string) => async (dispatch: Dispatch, getState: () => State) => {
+const addUserToThread = (displayName: string, emoji: string, endChatHandler: () => void) => async (dispatch: Dispatch, getState: () => State) => {
```
Comment (alkwa-msft) [inline Chat/ClientApp/src/core/sideEffects.ts:39]: kicked?

File: Chat/ClientApp/src/core/sideEffects.ts [hunk 10]
```diff
@@ -143,6 +142,74 @@ const subscribeForReadReceipt = async (
...
 };
 
+const subscribeForChatParticipants = async (chatClient: ChatClient, identity: string, endChatHandler: ()=> void, dispatch: Dispatch, getState: () => State) => {
+  chatClient.on('participantsRemoved', async (event: ParticipantsRemovedEvent) => {
+    const state = getState();
+    let participants: ChatParticipant[] = [];
+    for(let chatParticipant of event.participantsRemoved) {
+      // if you are in the list, remove yourself from the chat
+      if(isUserMatchingIdentity(chatParticipant.user, identity)) {
+        setThreadMembersError(true);
```
Comment (alkwa-msft) [inline Chat/ClientApp/src/core/sideEffects.ts:152]: verify if we need to set this error?

File: Chat/ClientApp/src/core/sideEffects.ts [hunk 11]
```diff
@@ -143,6 +142,74 @@ const subscribeForReadReceipt = async (
...
+        endChatHandler();
```
Comment (alkwa-msft) [inline Chat/ClientApp/src/core/sideEffects.ts:153]: kick yoself

File: Chat/ClientApp/src/core/sideEffects.ts [hunk 12]
```diff
@@ -143,6 +142,74 @@ const subscribeForReadReceipt = async (
...
+        return;
+      }
+    }
+
+    const originalParticipants = state.threadMembers.threadMembers;
+    for(var i = 0; i < originalParticipants.length; i++) {
+      if (event.participantsRemoved.filter(
```
Comment (alkwa-msft) [inline Chat/ClientApp/src/core/sideEffects.ts:160]: can we make this less wordy or add in a better comment?

File: Chat/ClientApp/src/core/sideEffects.ts [hunk 13]
```diff
@@ -251,12 +313,24 @@ const getMessages = () => async (dispatch: Dispatch, getState: () => State) => {
...
   }
-  let messages = await getMessagesHelper(await chatClient.getChatThreadClient(threadId), threadId);
+  let messages = await getMessagesHelper(await chatClient.getChatThreadClient(threadId));
   if (messages === undefined) {
     console.error('unable to get messages');
     return;
   }
-  return dispatch(setMessages(messages.reverse()));
+
+  const clientChatMessages = messages.map(message => { return {
```
Comment (alkwa-msft) [inline Chat/ClientApp/src/core/sideEffects.ts:322]: can we use the convertToClientChatMessage function here?

File: Chat/ClientApp/src/core/sideEffects.ts [hunk 14]
```diff
@@ -302,10 +376,14 @@ const removeThreadMemberByUserId = (userId: string) => async (dispatch: Dispatch
...
-  });
-  if (response._response.status === TOO_MANY_REQUESTS_STATUS_CODE) {
+  try {
+    let response = await chatThreadClient.removeParticipant({
+      communicationUserId: userId
+    });
+  }
+  catch(error) {
+    // TOO_MANY_REQUESTS_STATUS_CODE
+    console.log(error);
```
Comment (alkwa-msft) [inline Chat/ClientApp/src/core/sideEffects.ts:386]: can we test this?

File: Chat/ClientApp/src/core/sideEffects.ts [hunk 15]
```diff
@@ -433,7 +529,7 @@ const addThreadMemberHelper = async (threadId: string, user: User, dispatch: Dis
       body: JSON.stringify(body)
     };
     let response = await fetch('/addUser/' + threadId, addMemberRequestOptions);
-    dispatch(setAddThreadMemberError(response.status !== MULTI_STATUS));
+    dispatch(setAddThreadMemberError(response.status !== OK));
```
Comment (alkwa-msft) [inline Chat/ClientApp/src/core/sideEffects.ts:532]: is this going to be an issue?

File: Chat/ClientApp/src/core/sideEffects.ts [hunk 16]
```diff
@@ -442,7 +538,7 @@ const addThreadMemberHelper = async (threadId: string, user: User, dispatch: Dis
 const getThreadMembersHelper = async (chatThreadClient: ChatThreadClient) => {
```
Comment (alkwa-msft) [inline Chat/ClientApp/src/core/sideEffects.ts:538]: can we confirm if we really even get getThreadMembers as well?

File: Chat/ClientApp/src/core/sideEffects.ts [hunk 17]
```diff
@@ -471,105 +567,68 @@ const sendMessageHelper = async (
...
-            dispatch,
-            retryCount + 1,
-            getState
-          );
-        }, 200);
+
+    const messageResult = await chatThreadClient.sendMessage(SendMessageRequest);
+
+    if (messageResult.id) {
+      let message: ChatMessage | undefined = await getMessageHelper(chatThreadClient, messageResult.id);
```
Comment (alkwa-msft) [inline Chat/ClientApp/src/core/sideEffects.ts:574]: lets just double check this funtionality again

File: Chat/ClientApp/src/core/sideEffects.ts [hunk 18]
```diff
@@ -591,7 +650,7 @@ const getMessagesHelper = async (chatThreadClient: ChatThreadClient, threadId: s
       }
 
       // filter and only return top 100 text messages
-      messages.push(...messages_temp.filter((message) => message.type === 'Text'));
+      messages.push(...messages_temp.filter((message) => message.type === 'text'));
```
Comment (alkwa-msft) [inline Chat/ClientApp/src/core/sideEffects.ts:653]: We should add the correct type here if the enum changes later on

**Comment by JamesBurnside** [comment]
Found no new issues in the smoke test 👍👍
(One small improvement suggested and one existing bug found)

Tested:
* Creating new chat thread
* Sending messages
* Having another member join
* Kicking the other member
  * Other member saw the oops! you are no longer a participant screen correctly
  * remaining participants saw kicked member gone
* Leaving chat thread correctly removes member from the list
  * **improvement**: Member who left sees oops screen - I'd expect to not see the word "oops" from an intentional action
    
image

* Typing indicator works (although it's like 10 seconds before it disappears - not a regression in this PR though)
* Setting group name
* **Existing Bug**: Not a regression in this PR but there's a bug where the senders messages get out of order if sent quickly:
  
image

  (Left shows the sender, the order should match the bottom row of the keyboard - for the receiver on the right it does match, v. easy to repro)

File: Chat/ClientApp/src/core/sideEffects.ts (hunk 16 above)
Comment (alkwa-msft) [reply Chat/ClientApp/src/core/sideEffects.ts:538]: we only used this helper in one spot so I removed the function and just moved the lines over to its one usage area

File: Chat/ClientApp/src/core/sideEffects.ts (hunk 8 above)
Comment (alkwa-msft) [reply Chat/ClientApp/src/core/sideEffects.ts:17]: removing cooldown period for now

File: Chat/ClientApp/src/core/sideEffects.ts (hunk 13 above)
Comment (alkwa-msft) [reply Chat/ClientApp/src/core/sideEffects.ts:322]: yes using that now resolving this.

File: Chat/ClientApp/src/core/sideEffects.ts (hunk 14 above)
Comment (alkwa-msft) [reply Chat/ClientApp/src/core/sideEffects.ts:386]: I don't think this is going to be an issue anymore but we should still bug bash this with the team

**Pull request merged**